Mathlib/Data/Finset/Antidiagonal.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2023 Antoine Chambert-Loir and María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández, Bhavik Mehta
+-/
+
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Finsupp.Defs
+import Mathlib.Data.Finsupp.Interval
+
+/-! # Antidiagonal with values in general types
+
+Let `n : μ`, where `μ` is a canonically ordered add monoid with locally finite order.
+
+* For `s : Finset ι`, we define `Finset.Pi.antidiagonal s n` as the `Finset (ι → μ)`
+of functions with support in `s` whose sum is equal to `n`.
+
+* We define `Finset.antidiagonal n : Finset (μ × μ)` of pairs adding to `n`.
+
+These definitions generalize `Finset.Nat.antidiagonal` and `Finsupp.antidiagonal`
+which are defined in `Mathlib.Data.Finset.NatAntidiagonal`
+and `Mathlib.Data.Finsupp.Antidiagonal` and we make the comparisons there.
+
+This definition does not exactly match with that of `Multiset.antidiagonal` which is
+defined in `Mathlib.Data.Multiset.Antidiagonal`, because of the multiplicities.
+Indeed, by counting multiplicities, `Multiset α` is equivalent to `α →₀ ℕ`,
+but `Finset.antidiagonal` and `Multiset.antidiagonal` will return different objects.
+For example, for `s : Multiset ℕ := {0,0,0}`, `Multiset.antidiagonal s` has 8 elements
+but `Finset.antidiagonal s` has only 4.
+
+-- def s : Multiset ℕ := {0, 0, 0}
+-- #eval (Finset.antidiagonal s).card -- 4
+-- #eval Multiset.card (Multiset.antidiagonal s) -- 8
+
+This is mostly taken from a mathlib3 file of Bhavik Mehta,
+https://leanprover-community.github.io/mathlib_docs/wiedijk_100_theorems/partition.html
+-/
+
+open scoped BigOperators
+
+open Finset Function
+
+variable {μ : Type _} [CanonicallyOrderedAddMonoid μ] [LocallyFiniteOrder μ] [DecidableEq μ]
+
+variable {ι : Type _} [DecidableEq ι] [DecidableEq (ι → μ)]
+
+namespace Finset
+
+/-- The Finset of pairs of elements of `μ` with sum `n` -/
+def antidiagonal (n : μ) : Finset (μ × μ) :=
+  Finset.filter (fun uv => uv.fst + uv.snd = n) (Finset.product (Iic n) (Iic n))
```

Comment:
Should we have a `HasAntidiagonal` typeclass? That would allow us to have a O(n) rather than O(n^2) implementation, and I can see this being executed / reduced in combinatorics.

What types `μ` do we care about in practice?

Reply:
1. For the `HasAntidiagonal` typeclass. Yes, gladly, if that allows to have a faster implementation, but I don't know what to change exactly.

2. We will use for a type`μ` the type of finitely supported functions from some type to Nat.

Reply:
To be precise, we will use piAntidiagonal to express the coefficients of the products of power series and prove some results such as : if a power series as a nilpotent coefficient as constant term, then its powers converge to 0 in a strong sense (= simple convergence on coefficients, discrete topology on the target, meaning than each coefficient of f ^ n is eventually 0 when n goes to infinity)